cmake/modules/FindFmt.cmake

```diff
@@ -25,12 +25,23 @@ if(ENABLE_INTERNAL_FMT)
 
   set(FMT_VERSION ${${MODULE}_VER})
 
+  if(WIN32 OR WINDOWS_STORE)
+    # find the path to the patch executable
+    find_program(PATCH_EXECUTABLE NAMES patch patch.exe REQUIRED)
```

Comment:
patch.exe has been difficult on windows. I believe we actually ship a patched version ourselves to work around some of the issues with it. When possible, I try to use git patch instead to avoid any issues. 

It's not an issue as we use patch in more places, just something to keep in mind

Reply:
yeah, i wasnt sure how to handle it really. we have a couple of different things in our cmake. addons download presumably the patched version you speak of from our mirrors. https://github.com/xbmc/xbmc/blob/master/cmake/scripts/windows/tools/patch.cmake

Maybe i should pull that into a Find module, and then we can download if not available as part of the find_program call?

Or on windows, are we just better to add the patch executable as a requirement in build docs, and the dev is expected to provide a patch of some sort on their PATH?

Reply:
Seems like a can of worms, let's ignore it for now

Reply:
Happy to punt it down the road for a later time.

Just to give you an idea of what im playing with, im looking at adding ENABLE_INTERNAL_XXX for a few other deps that have a cmake build system upstream (taglib, pcre, mariadb-c-connector, libxml2, libnfs + existing ones) and get them to build internally for all platforms. They are all reasonably small/quick to build.

Ive got a wip setup for all tools/depends/target deps for unix platforms, but not sure how to handle autoconf based deps for windows just yet. I know people have tried a few different routes (eg ffmpeg and msys, rechi doing custom cmake scripts, etc). So i'll look into that down the line. Dont know if it'll go anywhere for windows, or just stick with your current system. But it gives me something to do for now, haha.

We've then got a couple deps using meson (dav1d, harfbuzz), so that'll be its own kettle of fish eventually as well, haha. Thinking that may just be best with adding a requirement to install meson via their msi

Reply:
hehe, autoconf stuff is a pain on windows, my personal feeling about them is I accept it for ffmpeg and emulators because they're too complex to write and maintain cmake stuff for, everything else we depend on have been fairly simple to replace with cmake when only caring about Windows 😄 

meson works well on windows but last I looked at it, it couldn't find tooling, so it requires some cmake wrappers to set things up before it can be invoked. From experience, dav1d is no fun at all to write a cmake build for, it uses a lot of macros and "weird" file includes to generate code for 8bit/16bit handling

Reply:
yeah, i have a feeling for most autoconf stuff on windows, a manual cmake file will be the easiest way to go. The only thing that annoys me about this, anyone updating deps really needs to be on the ball with changes of the dep. new source files, renamed source files, etc, to manually keep the cmakelists.txt in sync. A lot more effort required otherwise i can see issues cropping up because of a lazy dep update not being done right.

i think being able to rely on the upstream project to know their changes and keep their build system functional is higher than the average kodi dev looking to update a dep with a custom cmake build system. Bit of a catch 22. trying to make dep updates easier, ends up making them difficult on windows.

Im really hoping we could get it to a state that someone other than yourself is willing to actually do a windows dep update. The way i see it, we either need to integrate the kodi-deps repo into a jenkins job that can be updated, and auto uploaded to mirrors, make it so deps can be made/tested in tree (ala tools/depends, with jenkins building it and creating artifacts for testing) or maybe i should just give up the hope and let windows deps rot :P

Reply:
i guess another option is to script and rely on WSL/WSL2. but that brings its own large kettle of fish (ie limits building on win10+). I havent looked into WSL usage regarding autoconf/etc, but from what ive read its reasonably doable.
Problem as always though is someone doing the work that has the knowledge, haha.

Reply:
> meson works well on windows but last I looked at it, it couldn't find tooling, so it requires some cmake wrappers to set things up before it can be invoked.

What kind of tooling? e.g. the C/C++ compiler?

Reply:
Yeah, the Visual C++ compiler/linker.
looking at meson 0.64 it appears to find it when using `meson compile -C build` but not sure if it can handle cross compilation like cmake does it. When compiling for say arm64 windows store in cmake I run `cmake -G "Visual Studio 17 2022" -A arm64 -DCMAKE_SYSTEM_NAME=WindowsStore -DCMAKE_SYSTEM_VERSION=10.0` and cmake/msbuild will find the paths to the arm64 tooling, the latest sdk version and set things up correctly.

Meson appear to suggest running in a vs dev prompt https://mesonbuild.com/Using-with-Visual-Studio.html which is sort of ok when building a single project, it's a bit annoying when trying to integrate meson in another build like what we're trying to do.

I haven't investigated it thoroughly and I would be very happy to be proven wrong 😄

Reply:
I think the real problem here is probably the fact that Visual Studio doesn't really work unless you have a bunch of environment variables set from running the vcvars bat script.

Meson can try to parse that for you using `meson compile` instead of `ninja`, but running an extra program to parse vcvars and set variables, then run `cl.exe`, for *every* command in the Makefile/build.ninja, would be extremely slow and inefficient.

Does cmake really allow you to generate its Makefiles or Ninja backend together with the VS compiler, and *not* need to run it in a vs dev prompt? How curious.

...

On the topic of detecting the arm64 tooling when cross compiling, that is... a good point, and it doesn't seem Meson does this (it uses the current platform when determining which vcvars file to run) but I think that is probably just because the person who implemented the automatic vsenv handling simply didn't think of it. We can probably fix that. :)

Reply:
It seems to be something specific for the visual studio generator in cmake. When using ninja or makefiles it has to be configured manually or by running it in a visual studio dev prompt.